### PR TITLE
UI for creating/running VM program

### PR DESCRIPTION
**Pull request by tballmsft**

### CODE REVIEW

File: src/components/vm/VMRunner.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import { IT4Program } from "../../../jacdac-ts/src/vm/ir"
+import { IT4ProgramRunner } from "../../../jacdac-ts/src/vm/vmrunner"
+import { JDService } from "../../../jacdac-ts/src/jdom/service"
+import useChange from "../../jacdac/useChange"
+import DashboardDevice from "../dashboard/DashboardDevice"
+import LoadingProgress from "../ui/LoadingProgress"
+
+export default function VMRunner(props: {
+    json: IT4Program
+    roleManager: JDService
```
Comment (pelikhan) [inline src/components/vm/VMRunner.tsx:17]: you probably want to use the roleManager client from the bus. It already talks to the role manager to resolve the roles etc...

**Comment by pelikhan** [comment]
As discussed, let's decouple the role manager from the VM by introduction some kind of "service resolver" service that takes care of mapping names to device/service index (role maanger is an example of this).
For scripting purposes in the browser, a user should be able to write things like this:

- [ ] ``up`` , match any service that exposes a up event and use it (note: let's make sure our spec has unique event names)
- [ ] ``button.up``, match any button use up event
- [ ] ``A.up`` match any service whose instance name is "A" and pick the up event
- [ ] ``deviceid|shortid.up`` precise match on device id + find any event called up
- [ ] ``button.B.up``, match any button service, service instance "B", use up event

Down the road, would we support assignment instead of writeRegister? ```writeRegister(reg, val)`` => ``reg = val``

File: src/components/vm/VMRunner.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                    onClick={handleRun}
+                >
+                    Run
+                </Button>
+                <Button variant="outlined" onClick={handleCancel}>
+                    Cancel
+                </Button>
+            </Grid>
+            {roleManager && (
+                <Grid item xs={3}>
```
Comment (pelikhan) [inline src/components/vm/VMRunner.tsx:53]: use ``<Grid item xs>`` to allow the device element to use the remaining of the space

**Comment by tballmsft** [comment]
> Down the road, would we support assignment instead of writeRegister? ```writeRegister(reg, val)=>reg = val``

The current "syntax" is just a representation of abstract syntax that is conveniently supported by expression trees.  We can definitely have a different concrete syntax once we figure out what we want in the language. Of course, we might also have a UI for programming that doesn't use text...

**Comment by pelikhan** [comment]
Right we need structured editor of sorts

________________________________

**Pull request merged**